lib/src/main/java/com/dynatrace/metric/util/MetricLinePreConfiguration.java

```diff
@@ -0,0 +1,290 @@
+        return;
+      }
+
+      String normalizedKey = key;
+      if (DimensionKeyValidator.normalizationRequired(key)) {
+        NormalizationResult normalizationResult = Normalizer.normalizeDimensionKey(key);
+
+        normalizedKey = normalizationResult.getResult();
+        if (normalizationResult.messageType() != NormalizationResult.MessageType.NONE) {
+          logger.warning(() -> normalizationResult.getMessage());
+        }
+      }
+
+      // if the provided condition is met, the key-value would be overwritten during serialization,
+      // therefore, it can be already ignored at this time.
+      if (shouldBeIgnored.test(normalizedKey)) {
+        logger.info(
+            () ->
+                String.format(
+                    ValidationMessages.DIMENSION_DROPPED_KEY_OVERWRITTEN_MESSAGE, value, key));
+        return;
+      }
+
+      NormalizationResult normalizationResult =
+          Normalizer.normalizeDimensionValue(
+              value, MetricLineConstants.Limits.MAX_DIMENSION_VALUE_LENGTH);
+      if (normalizationResult.messageType() != NormalizationResult.MessageType.NONE) {
+        logger.warning(() -> normalizationResult.getMessage());
+      }
+
+      tryAddDimensionTo(normalizedKey, normalizationResult.getResult(), targetDimensions);
+    }
+
+    /**
+     * Attempts to append new dimension to the specified dimensions. It does so by checking if this
+     * additional dimension, would exceed the limit of {@value
+     * MetricLineConstants.Limits#MAX_DIMENSIONS_COUNT} dimensions. and if it will cause the metric
+     * line to exceed the {@value MetricLineConstants.Limits#MAX_LINE_LENGTH} length limit.
+     *
+     * @param normalizedKey The dimension key.
+     * @param normalizedValue The dimension value.
+     * @param targetDimensions The dimensions where key-value-pair should be added to
+     * @throws MetricException if appending the dimension will cause the metric line to exceed the
+     *     {@value MetricLineConstants.Limits#MAX_LINE_LENGTH} length limit, or it will exceed the
+     *     limit of {@value MetricLineConstants.Limits#MAX_DIMENSIONS_COUNT} dimensions.
+     */
+    private void tryAddDimensionTo(
+        String normalizedKey, String normalizedValue, Map<String, String> targetDimensions)
+        throws MetricException {
+      if (StringValueValidator.isNullOrEmpty(normalizedKey)) {
+        logger.warning(() -> ValidationMessages.DIMENSION_DROPPED_KEY_EMPTY_MESSAGE);
+        return;
+      }
+
+      if (this.dimensionCount + 1 > MetricLineConstants.Limits.MAX_DIMENSIONS_COUNT) {
+        throw new MetricException(ValidationMessages.TOO_MANY_DIMENSIONS_MESSAGE);
+      }
+
+      int dimSize =
+          normalizedKey.length()
+              + Character.charCount(CodePoints.EQUALS)
+              + normalizedValue.length()
+              + Character.charCount(CodePoints.COMMA);
+
+      if (this.serializationLength + dimSize > MetricLineConstants.Limits.MAX_LINE_LENGTH) {
```

Comment:
Todo: Ensure that we don't use that size limit in metadata, otherwise we might be throwing because we reached the limit here and then created metadata later. I think the two are separate,  but it might be good to double check

Reply:
Metadata lines do not have a size limit, as far as I know/understand

Reply:
I think they follow the same limits as the "normal" metric lines. It's impossible to create a metadata line that is too long though, I think, due to the restrictions for displayName/unit/description. I just wanted to make sure we don't throw when creating a metadata line just because the metric line is too long. However... that would probably also be okay, becuase whats the point of making a metadata line when the metric line is invalid anyway?

Reply:
I checked it, and we don't throw on the line being too long so consider this done